### PR TITLE
Add `TaskSeq.singleton` to the surface area

### PR DESCRIPTION
**Pull request by gusty**
Add `singleton` function.

### CODE REVIEW

File: src/FSharp.Control.TaskSeq/TaskSeq.fs [hunk 1]
```diff
@@ -15,6 +15,20 @@ module TaskSeq =
         for c: 'T in [] do
             yield c
     }
+    
+    let singleton (source: 'T) =
```
Comment (abelbraaksma) [inline src/FSharp.Control.TaskSeq/TaskSeq.fs:19]: Please move it to `TaskSeqInternal` and call that function from here.

**Review by abelbraaksma — CHANGES_REQUESTED** [review]
This function will need some tests. See the approach I took for `empty`. It won't really fit the auto-generated tests, but we should at least have coverage of the three basic cases: before/middle/after iteration and that it behaves stable with a few minimal combined tests.

I can set this up if you like.

File: src/FSharp.Control.TaskSeq/TaskSeq.fs [hunk 2]
```diff
@@ -15,6 +15,20 @@ module TaskSeq =
...
+        { new IAsyncEnumerable<'T> with
+            member _.GetAsyncEnumerator(_) =
+                let mutable started = false
+                { new IAsyncEnumerator<'T> with
+                    member _.MoveNextAsync () =
+                        let r = ValueTask.FromResult (not started)
+                        started <- true
+                        r                        
+                    member _.get_Current () : 'T = if started then source else invalidOp "Enumeration has not started. Call MoveNextAsync."
```
[the remark this answers is not in the document]
Comment (gusty) [reply src/FSharp.Control.TaskSeq/TaskSeq.fs:28]: Why? Is there a contract that specify that?
Getting the current within an invalid state is an undefined operation, therefore invalid operation.
Although there's no contract, this is exactly what `seq` does.

File: src/FSharp.Control.TaskSeq/TaskSeq.fs (hunk 1 above)
Comment (gusty) [reply src/FSharp.Control.TaskSeq/TaskSeq.fs:19]: Not clear to me this request.
The idea is that `singleton` is a public function.
Why do you want to move it to an internal namespace?
Is it to re-use it within other internal functions?

Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fs:19]: Yes. It’s nothing more than organisational. Also, we group like functions. Say we add `singletonAsync`, we’d process both in a single function in `TaskSeqInternal`.

This division is slightly arbitrary, but I like to keep the larger details separate from the lean and mean implementation file. In the end, there’s no functional reason other than organisation and style.

**Comment by abelbraaksma** [comment]
There's a compile error now, not sure how rebase introduced that, though:

> Error: D:\a\FSharp.Control.TaskSeq\FSharp.Control.TaskSeq\src\FSharp.Control.TaskSeq\TaskSeq.fs(25,43): error FS0039: The type 'ValueTask' does not define the field, constructor or member 'FromResult'.

File: src/FSharp.Control.TaskSeq/TaskSeq.fs (hunk 1 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fs:19]: Btw, just do be clear, the public function remains in `TaskSeq.fs`. That file should contain all surface area public functions. Only the implementation moves. Sorry for being vague earlier.

File: src/FSharp.Control.TaskSeq/TaskSeq.fs (hunk 2 above)
[the remark this answers is not in the document]
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fs:28]: After extensive research, the basic rule is: there is no rule. Both F# and BCL libs are inconsistent, sometimes throwing exceptions, sometimes not. We may throw exceptions at some point, but right now I prefer to be consistent with the CE. That is, I want this:

```F#
> let s: seq<int> = seq { do () } ;;
val s: seq<int>

> s.GetEnumerator().Current;;
val it: int = 0
```

to behave equal to this:

```f#
> let s: seq<int> = Seq.empty;;
val s: seq<int>

> s.GetEnumerator().Current;;
System.InvalidOperationException:  // this is not consistent in Seq
```

In other words: we will *not* throw an exception here for `taskSeq`. I may change this resolution, but for now, I think consistency is better, even though I agree in principal that throwing an exception isn't bad per se.

**Comment by abelbraaksma** [comment]
This should fix the compile error.

File: src/FSharp.Control.TaskSeq/TaskSeq.fs (hunk 1 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fs]: This is now done. I've also fixed the `singleton` implementation to properly return `defaultof` whenever before/after iteration.

**Review by abelbraaksma — APPROVED** [review]

**Comment by abelbraaksma** [comment]
Thanks for this! Merged into `main` now.

**Pull request merged**